rust/src/client.rs

```diff
@@ -0,0 +1,842 @@
+            settings_write_res_e_SETTINGS_WR_SETTING_REJECTED => WriteSettingError::SettingRejected,
+            settings_write_res_e_SETTINGS_WR_PARSE_FAILED => WriteSettingError::ParseFailed,
+            settings_write_res_e_SETTINGS_WR_READ_ONLY => WriteSettingError::ReadOnly,
+            settings_write_res_e_SETTINGS_WR_MODIFY_DISABLED => WriteSettingError::ModifyDisabled,
+            settings_write_res_e_SETTINGS_WR_SERVICE_FAILED => WriteSettingError::ServiceFailed,
+            settings_write_res_e_SETTINGS_WR_TIMEOUT => WriteSettingError::Timeout,
+            _ => WriteSettingError::Unknown,
+        }
+    }
+}
+
+impl std::fmt::Display for WriteSettingError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            WriteSettingError::ValueRejected => {
+                write!(f, "setting value invalid",)
+            }
+            WriteSettingError::SettingRejected => {
+                write!(f, "setting does not exist")
+            }
+            WriteSettingError::ParseFailed => {
+                write!(f, "could not parse setting value ")
+            }
+            WriteSettingError::ReadOnly => {
+                write!(f, "setting is read only")
+            }
+            WriteSettingError::ModifyDisabled => {
+                write!(f, "setting is not modifiable")
+            }
+            WriteSettingError::ServiceFailed => write!(f, "system failure during setting"),
+            WriteSettingError::Timeout => write!(f, "request wasn't replied in time"),
+            WriteSettingError::Unknown => {
+                write!(f, "unknown settings write response")
+            }
+        }
+    }
+}
+
+impl std::error::Error for WriteSettingError {}
+
+#[derive(Debug, Clone, Copy)]
+pub struct ReadSettingError {
+    code: u32,
+}
+
+impl From<i32> for ReadSettingError {
+    fn from(code: i32) -> Self {
+        (code as u32).into()
+    }
+}
+
+impl From<u32> for ReadSettingError {
+    fn from(code: u32) -> Self {
+        Self { code }
+    }
+}
+
+impl std::fmt::Display for ReadSettingError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "settings read failed with status code {}", self.code)
+    }
+}
+
+impl std::error::Error for ReadSettingError {}
+
+struct ClientInner<'a> {
+    context: *mut Context<'a>,
+    api: *mut settings_api_t,
+    ctx: *mut settings_t,
+}
+
+impl Drop for ClientInner<'_> {
+    fn drop(&mut self) {
+        // Safety: These were created via Box::into_raw and are not
+        // freed by the c library. We don't need to free event because
+        // it is pointer to the event held inside context so it gets dropped
+        unsafe {
+            let _ = Box::from_raw(self.context);
+            let _ = Box::from_raw(self.api);
+            settings_destroy(&mut self.ctx);
+        }
+    }
+}
+
+unsafe impl Send for ClientInner<'_> {}
+unsafe impl Sync for ClientInner<'_> {}
```

Comment:
For reading order, it would be nice if these came immediately before or after the `struct Client` definition